src/libfsm/print/c.c

```diff
@@ -478,6 +484,41 @@ fsm_print_c_body(FILE *f, const struct ir *ir,
 		}
 	}
 
+	/* This flag indicates whether the any of the input stream was
+	 * consumed before getting EOF and skipping over the state and
+	 * character logic expanded here.
+	 *
+	 * lx needs to track this for proper EOF handling. It previously
+	 * generated the state enum itself, so that it could include an
+	 * additional 'NONE' state. Inside the input loop, the default
+	 * state of NONE would be updated to the start state, but if the
+	 * input loop was skipped it would still be NONE. */
+	fprintf(f, "\tint has_consumed_input = 0;\n");
+
+	/* For FSM_IO_STR and FSM_IO_PAIR, define a macro that will be
```

Comment:
This macro is a bit ugly, but much simpler than changing normal code generation to instantiate some kind of character iterator, just so lx and libfsm are on the same page about advancing and pushing back the character stream.

Reply:
This macro rubs me the wrong way too. There's gotta be a nicer way. Maybe we could have lx emit a call to `%sadvance_end()` somewhere (in a new hook, e.g. at the end of the generated fsm loop) instead.